LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree structure instead of parent references.

OrdiNeu opened the pull request:
[Jira link](https://phenotips.atlassian.net/browse/LFS-911)

This mostly changes the backend for all subjects and subjectTypes. You can see the difference in `composum`, as parentage is now displayed as a tree structure, like below:

![image](https://user-images.githubusercontent.com/4656440/108799723-015ac180-755f-11eb-8283-9758612d618b.png)

**To test:** Just need to verify that this change didn't break anything -- the biggest changes have affected the creation of subjects and subject types.

**TO DO**:
- [x] upload variants page displays a 404 error
- [x] deleting subjects and subject types doesn't work
- [x] uploading variants doesn't work

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

[wip] Move the tumor/patient structure around to tree structures.

OrdiNeu committed:
[wip] Refactor work

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

Fix the subject selector

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

Move the subject creation to children instead of parent references.

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

Add the Progeny editor to handle creating progeny/parent/ancestor
references

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

Correct the subject types and remove some debugging output

marta- commented on lfs-resources/clinical-data/src/main/resources/SLING-INF/content/SubjectTypes/Patient/Tumor.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "jcr:primaryType": "lfs:SubjectType",
+  "label": "Tumor",
+  "lfs:defaultOrder": 1,
+  "TumorRegion": {
+    "jcr:primaryType": "lfs:SubjectType",
+    "label": "Tumor region",
+    "lfs:defaultOrder": 2
+  }
```
Why not keep Tumor and TumorRegion separate?

sdumitriu replied:
Um, we may have a problem. Since now Tumor is a child of Patient, but they are defined in different modules, when upgrading the module with Patient, it will erase the other two. We need to make sure that either we always import data-entry then lfs-resources, or Patient will only overwrite its properties, but not its child nodes.

OrdiNeu replied:
The overwrite is disabled in `clinical-data/pom.xml`, but I've also added the overwriteProperties as per standup in `
f3bde24` :heavy_check_mark:

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectSelector.jsx:
```diff
@@ -386,17 +387,33 @@ export function NewSubjectDialog (props) {
+    let newAllowedTypeParent = type["@path"].split("/").slice(0, -1).join("/");
+    let promise;
+    if (newAllowedTypeParent) {
+      promise = fetch(`${newAllowedTypeParent}.full.json`)
```
No `WithReLogin` here?

OrdiNeu replied:
Fixed in `0efb7dee` :heavy_check_mark:

marta- reviewed (commented):
The only thing that seems to be broken so far is the SubjectTypes admin page, which no longer shows subject types with parents.

acrowthe commented:
Found an issue:
- Subjects table on the dashboard does not show existing Tumor and Tumor Region subject, but does show existing Patients 
![image](https://user-images.githubusercontent.com/20056751/109024396-b7bbc500-768b-11eb-9193-deabf792d7bc.png)
![image](https://user-images.githubusercontent.com/20056751/109024425-c0140000-768b-11eb-921d-e51289dd6d0e.png)

marta- commented:
> Found an issue:
> 
> * Subjects table on the dashboard does not show existing Tumor and Tumor Region subject, but does show existing Patients

Right, all the tables that display subjects now need to be adapted to the new paths.

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

Allow child subjects in the pagination servlet

OrdiNeu commented:
@acrowthe 
> Found an issue:
> 
>     * Subjects table on the dashboard does not show existing Tumor and Tumor Region subject, but does show existing Patients

Fixed in `188f263` :heavy_check_mark:

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

Allow the clinical-data module to overwrite properties.

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

Replace fetch() calls with fetchWithReLogin() calls.

marta- commented on lfs-resources/clinical-data/pom.xml:
```diff
@@ -44,7 +44,7 @@
               SLING-INF/content/libs/lfs/resources/;path:=/libs/lfs/resources/;overwrite:=true,
               SLING-INF/content/Questionnaires/;path:=/Questionnaires/;overwrite:=true;checkin:=true,
-              SLING-INF/content/SubjectTypes/;path:=/SubjectTypes/;overwrite:=true
+              SLING-INF/content/SubjectTypes/;path:=/SubjectTypes/;overwrite:=false,
```
does this need `,overwriteProperties:=true`?

OrdiNeu replied:
It doesn't currently, but in the future if a node is redefined in the clinical-data module, it should have precedence over the defaults.

marta- reviewed (changes requested):
Tumor and Tumor region need to be declared in separate JSONs @OrdiNeu @sdumitriu

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

Split Tumor.json into Tumor.json and TumorRegion.json

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

Fix documentation in PaginationServlet.java

OrdiNeu commented:
@marta- 
> Tumor and Tumor region need to be declared in separate JSONs 

Fixed in `e23333db` :heavy_check_mark: 

 @sdumitriu
>Now the comment above is no longer accurate... 

Fixed in `e2175e4` :heavy_check_mark:

marta- commented on lfs-resources/clinical-data/src/main/resources/SLING-INF/content/SubjectTypes/Patient/Tumor.json:
```diff
@@ -1,6 +1,5 @@
-  "jcr:reference:parent": "../Patient",
   "lfs:defaultOrder": 1
-}
+}
```
I think we're supposed to have newlines at the end of files. Same for Tumor region.

marta- reviewed (dismissed):
Looks good to me.

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

Add newlines to end of file

sashaandjic commented:
![image](https://user-images.githubusercontent.com/23243244/109192193-efe30680-7764-11eb-8ee6-c2e2e4c909aa.png)

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

Fix the variant page

sashaandjic commented:
Variants upload fails 
![image](https://user-images.githubusercontent.com/23243244/109304536-6e45b400-780a-11eb-90d9-76a2bb0800ca.png)

sashaandjic commented:
What is the difference between `parents `and `ancestors`?
Why do we use `parents`, not `parent`, as we only show the first direct ancestor here?
![image](https://user-images.githubusercontent.com/23243244/109305555-e82a6d00-780b-11eb-9650-821df8ad4895.png)

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

Remove versionable nodes by checking them out, and remove only parents
to prevent version clashes.

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

Fix an issue with the deletion not saving properly.

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

Delay checkin of parent until after session save.
Remove debugging code.

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

Fix an issue with the variant uploader not using the proper subject
types.

OrdiNeu commented:
@sashaandjic 
> Variants upload fails

Fixed in `f5c1393` :heavy_check_mark:

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectSelector.jsx:
```diff
@@ -333,8 +334,8 @@ export function NewSubjectDialog (props) {
     // Grab the parent as an array if it exists, or the callback from the previously created parent, or use an empty array
-    let parent = newSubjectParent[index]?.["jcr:uuid"] || subject;
-    parent = (parent ? [parent] : []);
+    //let parent = newSubjectParent[index]?.["jcr:uuid"] || subject;
```
Remove

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

Remove debugging code.

sashaandjic commented:
patient link on tumor chart does not work
![image](https://user-images.githubusercontent.com/23243244/109998319-1cf06580-7cdf-11eb-94a5-eeb79e17d194.png)

sdumitriu commented on modules/subjects/src/main/java/ca/sickkids/ccm/lfs/subjects/internal/SubjectProgenyEditorProvider.java:
```diff
@@ -0,0 +1,60 @@
+ * @version $Id$
+ */
+@Component(name = "SubjectProgenyEditorProvider", service = EditorProvider.class,
+    scope = ServiceScope.SINGLETON, immediate = true)
```
```suggestion
@Component(immediate = true)
```

sdumitriu commented on modules/subjects/src/main/java/ca/sickkids/ccm/lfs/subjects/internal/SubjectProgenyEditor.java:
```diff
@@ -0,0 +1,218 @@
+
+/**
+ * An {@link Editor} that computes and sets the fullIdentifier property for every changed Subject. The full identifier
+ * is a concatenation of the subject's hierarchy, in the format {@code parent's full identifier / subject's identifier}.
```
Copy/paste leftover.

sdumitriu commented on modules/subjects/src/main/java/ca/sickkids/ccm/lfs/subjects/internal/SubjectProgenyEditor.java:
```diff
@@ -0,0 +1,218 @@
+
+    private static final String PROP_ANCESTORS = "ancestors";
+
+    private static final String PROP_PARENTS = "parents";
```
Is this still needed?

sdumitriu commented on modules/subjects/src/main/java/ca/sickkids/ccm/lfs/subjects/internal/SubjectProgenyEditor.java:
```diff
@@ -0,0 +1,218 @@
+    @Override
+    public void leave(NodeState before, NodeState after) throws CommitFailedException
+    {
+        //if (isRoot(this.currentNodeBuilder)) {
```
Remove.

sdumitriu commented on modules/subjects/src/main/java/ca/sickkids/ccm/lfs/subjects/internal/SubjectProgenyEditor.java:
```diff
@@ -0,0 +1,218 @@
+                computeProgeny();
+            } catch (RepositoryException e) {
+                // This is not a fatal error, the subject status is not required for a functional application
+                LOGGER.warn("Unexpected exception while computing the full identifier of subject {}",
```
c/p leftover

sdumitriu commented on modules/subjects/src/main/java/ca/sickkids/ccm/lfs/subjects/internal/SubjectProgenyEditor.java:
```diff
@@ -0,0 +1,218 @@
+                computeParents();
+                computeProgeny();
+            } catch (RepositoryException e) {
+                // This is not a fatal error, the subject status is not required for a functional application
```
```suggestion
                // This is not a fatal error, the subject progeny is not required for a functional application
```

sdumitriu commented on modules/subjects/src/main/java/ca/sickkids/ccm/lfs/subjects/internal/SubjectProgenyEditor.java:
```diff
@@ -0,0 +1,218 @@
+    }
+
+    /**
+     * Gather all identifiers from all the Subject parents of the current node and store them as the ' / ' separated
+     * fullIdentifier property.
```
c/p leftover

sdumitriu commented on modules/subjects/src/main/java/ca/sickkids/ccm/lfs/subjects/internal/SubjectProgenyEditor.java:
```diff
@@ -0,0 +1,218 @@
+     * @param nodeBuilder the current node
+     * @param session the session used to retrieve subjects by UUID
+     */
+    public SubjectProgenyEditor(final NodeBuilder nodeBuilder, final Session session)
```
This computes both progeny and ancestry, so either change the name (`SubjectPedigreeEditor`?), or split into two different editors.

sdumitriu commented on modules/subjects/src/main/java/ca/sickkids/ccm/lfs/subjects/internal/SubjectProgenyEditor.java:
```diff
@@ -0,0 +1,218 @@
+     * @param node the JCR Node to check
+     * @return {@code true} if the node is a Subject node, {@code false} otherwise
+     */
+    private boolean isRoot(NodeBuilder node)
```
This seems to be no longer in use, delete?

sdumitriu reviewed (changes requested):
Two bugs:
- Subjects created through the Variants Uploader are not properly nested, they are all children of /Subjects. This is not a critical bug, but would be nice to have it fixed.
- Recursive deletion needs to also check for referenced nodes for each child: create a Chemotherapy form for a Tumor, then try to delete the parent Patient.

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

Split the progeny editor into separate Ancestry/Progeny editors, with updated docstrings.

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

When deleting children, also check for their references.

OrdiNeu commented:
@sdumitriu 
> Recursive deletion needs to also check for referenced nodes for each child: create a Chemotherapy form for a Tumor, then try to delete the parent Patient. 

Fixed in `9e21cbc` :heavy_check_mark:

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

Fix references in the variant files container.

OrdiNeu commented:
@sdumitriu 
>     * Subjects created through the Variants Uploader are not properly nested, they are all children of /Subjects. This is not a critical bug, but would be nice to have it fixed.

Fixed in `7459bbf`

sdumitriu reviewed (changes requested):
Upload a file called `patient_tumor.csv`: exception, `file.region is undefined`.

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

Fix a whitescreen error when uploading a variants file without a region
specified.

sdumitriu commented:
Oh, not quite, it's a bit more complicated, let me figure out the steps.

sdumitriu commented:
Bug 1: Upload `p_t.csv`, then `p_t_r.csv` -> JS error.

Bug 2: Upload `p_t.csv` again -> new form has no subject.

marta- commented:
If the only remaining issues are strictly related to the Variants page, could we merge everything else to get the other PRs moving, and fix variants in a separate PR? @sdumitriu

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

Fix an issue with the variants uploader where the path would be
double-nested.

OrdiNeu committed:
LFS-911: Refactor lfs:Subject and lfs:SubjectType nodes to use a tree
structure instead of parent references.

Fix an error that occurs when uploading a file for an already-existing
tumour/region in the variants.

OrdiNeu commented:
@sdumitriu 
> Bug 1: Upload `p_t.csv`, then `p_t_r.csv` -> JS error. 

Fixed in `e3b8ca4` :heavy_check_mark: 

> Bug 2: Upload `p_t.csv` again -> new form has no subject. 

Fixed in `c850047` :heavy_check_mark:
